// @prisma/language-server/dist/src/test/hover.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const MessageHandler_1 = require("../MessageHandler");
const assert = __importStar(require("assert"));
const helper_1 = require("./helper");
function assertHover(position, expected, fixturePath) {
    const document = helper_1.getTextDocument(fixturePath);
    const params = {
        textDocument: document,
        position: position,
    };
    const hoverResult = MessageHandler_1.handleHoverRequest(document, params);
    assert.ok(hoverResult !== undefined);
    assert.deepEqual(hoverResult.contents, expected.contents);
    assert.deepEqual(hoverResult.range, expected.range);
}
suite('Hover of /// documentation comments', () => {
    const fixturePath = './hover.prisma';
    test('Model', () => {
        assertHover({
            character: 10,
            line: 23,
        }, {
            contents: 'Post including an author and content.',
        }, fixturePath);
    });
    test('Enum', () => {
        assertHover({
            character: 17,
            line: 24,
        }, {
            contents: 'This is an enum specifying the UserName.',
        }, fixturePath);
    });
});
