// @prisma/language-server/dist/src/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startServer = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
const checkpoint_client_1 = require("checkpoint-client");
const telemetry_1 = require("./telemetry");
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const MessageHandler = __importStar(require("./MessageHandler"));
const util = __importStar(require("./util"));
const install_1 = __importDefault(require("./install"));
const packageJson = require('../../package.json'); // eslint-disable-line
function getConnection(options) {
    let connection = options === null || options === void 0 ? void 0 : options.connection;
    if (!connection) {
        connection = process.argv.includes('--stdio')
            ? vscode_languageserver_1.createConnection(process.stdin, process.stdout)
            : vscode_languageserver_1.createConnection(new vscode_languageserver_1.IPCMessageReader(process), new vscode_languageserver_1.IPCMessageWriter(process));
    }
    return connection;
}
/**
 * Starts the language server.
 *
 * @param options Options to customize behavior
 */
function startServer(options) {
    const connection = getConnection(options);
    const documents = new vscode_languageserver_1.TextDocuments(vscode_languageserver_textdocument_1.TextDocument);
    // Does the clients accepts diagnostics with related information?
    let hasCodeActionLiteralsCapability = false;
    connection.onInitialize((params) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        telemetry_1.initializeTelemetry(connection);
        const capabilities = params.capabilities;
        hasCodeActionLiteralsCapability = Boolean((_b = (_a = capabilities === null || capabilities === void 0 ? void 0 : capabilities.textDocument) === null || _a === void 0 ? void 0 : _a.codeAction) === null || _b === void 0 ? void 0 : _b.codeActionLiteralSupport);
        const binPathPrismaFmt = yield util.getBinPath();
        if (yield util.binaryIsNeeded(binPathPrismaFmt)) {
            try {
                yield install_1.default(binPathPrismaFmt);
                connection.console.info(`Prisma plugin prisma-fmt installation succeeded.`);
            }
            catch (err) {
                telemetry_1.sendException(yield checkpoint_client_1.getSignature(), err, `Cannot install prisma-fmt.`);
                connection.console.error('Cannot install prisma-fmt: ' + err); // eslint-disable-line @typescript-eslint/restrict-plus-operands
            }
        }
        connection.console.info(`Installed version of Prisma binary 'prisma-fmt': ${util.getVersion()}`);
        connection.console.info(
        // eslint-disable-next-line
        `Extension name ${packageJson.name} with version ${packageJson.version}`);
        const prismaCLIVersion = util.getCLIVersion();
        connection.console.info(`Prisma CLI version: ${prismaCLIVersion}`);
        const result = {
            capabilities: {
                definitionProvider: true,
                documentFormattingProvider: true,
                completionProvider: {
                    resolveProvider: true,
                    triggerCharacters: ['@', '"', '.'],
                },
                hoverProvider: true,
                renameProvider: true,
            },
        };
        if (hasCodeActionLiteralsCapability) {
            result.capabilities.codeActionProvider = {
                codeActionKinds: [vscode_languageserver_1.CodeActionKind.QuickFix],
            };
        }
        return result;
    }));
    function validateTextDocument(textDocument) {
        return __awaiter(this, void 0, void 0, function* () {
            const diagnostics = yield MessageHandler.handleDiagnosticsRequest(textDocument, (errorMessage) => {
                connection.window.showErrorMessage(errorMessage);
            });
            connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
        });
    }
    documents.onDidChangeContent((change) => __awaiter(this, void 0, void 0, function* () {
        yield validateTextDocument(change.document);
    }));
    documents.onDidOpen((open) => __awaiter(this, void 0, void 0, function* () {
        yield validateTextDocument(open.document);
    }));
    function getDocument(uri) {
        return documents.get(uri);
    }
    connection.onDefinition((params) => __awaiter(this, void 0, void 0, function* () {
        const doc = getDocument(params.textDocument.uri);
        if (doc) {
            const definition = MessageHandler.handleDefinitionRequest(doc, params);
            if (definition !== undefined) {
                telemetry_1.sendTelemetry({
                    action: 'definition',
                    attributes: {
                        signature: yield checkpoint_client_1.getSignature(),
                    },
                });
            }
            return definition;
        }
    }));
    connection.onCompletion((params) => __awaiter(this, void 0, void 0, function* () {
        const doc = getDocument(params.textDocument.uri);
        if (doc) {
            return yield MessageHandler.handleCompletionRequest(params, doc);
        }
    }));
    connection.onCompletionResolve((completionItem) => __awaiter(this, void 0, void 0, function* () {
        telemetry_1.sendTelemetry({
            action: 'resolveCompletion',
            attributes: {
                label: completionItem.label,
                signature: yield checkpoint_client_1.getSignature(),
            },
        });
        return MessageHandler.handleCompletionResolveRequest(completionItem);
    }));
    connection.onDidChangeWatchedFiles(() => {
        // Monitored files have changed in VS Code
        connection.console.log(`Types have changed. Sending request to restart TS Language Server.`);
        // Restart TS Language Server
        connection.sendNotification('prisma/didChangeWatchedFiles', {});
    });
    connection.onHover((params) => __awaiter(this, void 0, void 0, function* () {
        const doc = getDocument(params.textDocument.uri);
        if (doc) {
            const hover = MessageHandler.handleHoverRequest(doc, params);
            if (hover !== undefined) {
                telemetry_1.sendTelemetry({
                    action: 'hover',
                    attributes: {
                        signature: yield checkpoint_client_1.getSignature(),
                    },
                });
            }
            return hover;
        }
    }));
    connection.onDocumentFormatting((params) => __awaiter(this, void 0, void 0, function* () {
        const doc = getDocument(params.textDocument.uri);
        if (doc) {
            telemetry_1.sendTelemetry({
                action: 'format',
                attributes: {
                    signature: yield checkpoint_client_1.getSignature(),
                },
            });
            return MessageHandler.handleDocumentFormatting(params, doc, (errorMessage) => {
                connection.window.showErrorMessage(errorMessage);
            });
        }
    }));
    connection.onCodeAction((params) => __awaiter(this, void 0, void 0, function* () {
        const doc = getDocument(params.textDocument.uri);
        if (doc) {
            const codeActions = MessageHandler.handleCodeActions(params, doc);
            if (codeActions.length !== 0) {
                telemetry_1.sendTelemetry({
                    action: 'codeAction',
                    attributes: {
                        signature: yield checkpoint_client_1.getSignature(),
                    },
                });
            }
            return codeActions;
        }
    }));
    connection.onRenameRequest((params) => __awaiter(this, void 0, void 0, function* () {
        const doc = getDocument(params.textDocument.uri);
        if (doc) {
            const rename = MessageHandler.handleRenameRequest(params, doc);
            if (rename !== undefined) {
                telemetry_1.sendTelemetry({
                    action: 'rename',
                    attributes: {
                        signature: yield checkpoint_client_1.getSignature(),
                    },
                });
            }
            return rename;
        }
    }));
    // Make the text document manager listen on the connection
    // for open, change and close text document events
    documents.listen(connection);
    connection.listen();
}
exports.startServer = startServer;
